server/src/main/java/org/elasticsearch/action/search/CCSSingleCoordinatorSearchProgressListener.java

```diff
@@ -0,0 +1,318 @@
+
+        for (Map.Entry<String, Integer> entry : totalByClusterAlias.entrySet()) {
+            String clusterAlias = entry.getKey();
+            int successfulCount = entry.getValue().intValue();
+
+            AtomicReference<SearchResponse.Cluster> clusterRef = clusters.getCluster(clusterAlias);
+            boolean swapped;
+            do {
+                SearchResponse.Cluster curr = clusterRef.get();
+                SearchResponse.Cluster.Status status = curr.getStatus();
+                if (status != SearchResponse.Cluster.Status.RUNNING) {
+                    // don't swap in a new Cluster if the final state has already been set
+                    break;
+                }
+                TimeValue took = new TimeValue(timeProvider.buildTookInMillis());
+                int successfulShards = successfulCount + curr.getSkippedShards();
+                assert successfulShards + curr.getFailedShards() == curr.getTotalShards()
+                    : "successfulShards("
+                        + successfulShards
+                        + ") + failedShards("
+                        + curr.getFailedShards()
+                        + ") != totalShards ("
+                        + curr.getTotalShards()
+                        + ')';
+                if (curr.isTimedOut() || successfulShards < curr.getTotalShards()) {
+                    status = SearchResponse.Cluster.Status.PARTIAL;
+                } else {
+                    assert successfulShards == curr.getTotalShards()
+                        : "successful (" + successfulShards + ") should equal total(" + curr.getTotalShards() + ") if get here";
+                    status = SearchResponse.Cluster.Status.SUCCESSFUL;
+                }
+                SearchResponse.Cluster updated = new SearchResponse.Cluster.Builder(curr).setStatus(status)
+                    .setSuccessfulShards(successfulShards)
+                    .setTook(took)
+                    .build();
+                swapped = clusterRef.compareAndSet(curr, updated);
+            } while (swapped == false);
+        }
+    }
+
+    /**
+     * Executed when a shard returns a fetch result.
+     *
+     * @param shardIndex The index of the shard in the list provided by {@link SearchProgressListener#onListShards})}.
+     */
+    @Override
+    public void onFetchResult(int shardIndex) {}
+
+    /**
+     * Executed when a shard reports a fetch failure.
+     *
+     * @param shardIndex The index of the shard in the list provided by {@link SearchProgressListener#onListShards})}.
+     * @param shardTarget The last shard target that thrown an exception.
+     * @param exc The cause of the failure.
+     */
+    @Override
+    public void onFetchFailure(int shardIndex, SearchShardTarget shardTarget, Exception exc) {}
+
+    private Map<String, Integer> partitionCountsByClusterAlias(List<SearchShard> shards) {
+        return shards.stream().collect(Collectors.groupingBy(shard -> {
```

Comment:
You are mutating the map returned here in at least one spot when you merge in the skipped counts.
We shouldn't be doing that, the docs for the collector state:

```
There are no guarantees on the type, mutability, serializability, or thread-safety of the Map returned.
```

We should make sure to get a mutable map here somehow (or really just use a normal loop and collect to a `HashMap` :).

Reply:
Thanks for the info on the Collector state.

I have now changed `partitionCountsByClusterAlias` to accept a stream, so for the case where I need both lists, I do:

```java
        Map<String, Integer> totalByClusterAlias = partitionCountsByClusterAlias(Stream.concat(shards.stream(), skipped.stream()));
```